index.d.ts

```diff
@@ -0,0 +1,597 @@
+// TypeScript Version: 4.0
+
+// Note: this is a `.d.ts` file because it is not possible to have default type
+// parameters in JSDoc-based TypeScript, which is a feature we use to type that:
+//
+// ```js
+// .use(somePlugin, theOptions)
+// ```
+//
+// `theOptions` matches the options that `somePlugin` expects and thus is very
+// important for making unified usable in TypeScript.
+//
+// Furthermore, this is places in the root of the project because types that
+// accept type parameters cannot be re-exported as such easily.
+
+import {Node} from 'unist'
+import {VFile, VFileCompatible} from 'vfile'
+
+/**
+ * Processor allows plugins to be chained together to transform content.
+ * The chain of plugins defines how content flows through it.
+ */
+export interface Processor extends FrozenProcessor {
+  /**
+   * Configure the processor to use a plugin.
+   *
+   * @typeParam S
+   *   Plugin settings.
+   * @param plugin
+   *   Plugin (function) to use.
+   *   Plugins are deduped based on identity: passing a function in twice will
+   *   cause it to run only once.
+   * @param settings
+   *   Configuration for plugin, optional.
+   *   Plugins typically receive one options object, but could receive other and
+   *   more values.
+   *   It’s also possible to pass a boolean instead of settings: `true` (to turn
+   *   a plugin on) or `false` (to turn a plugin off).
+   * @returns
+   *   Current processor.
+   */
+  use<S extends any[] = any[]>(
```

Comment:
(optional) it could be good time to give `S` a more descriptive name
```suggestion
  use<Settings extends any[] = any[]>(
```

Reply:
I'm cautious about defaulting to allow any number of settings of any type by default.
It may be a good default to start with no settings, and let plugins add them as needed
```suggestion
  use<S extends any[] = []>(
```

Reply:
Good idea to make it more descriptive. Settings is also a bit vague though. I think especially the fact that it's an array, for the variadic args, is confusing and doesn't work well with Settings. What about 'PluginArguments'? 'ExpectedParameters'?

Reply:
I'd lean towards `PluginParameters` or `PluginArguments`

Reply:
On `any[]` => `[]`:

1. xo through `@typescript-eslint/ban-types` doesn’t allow `[]`.
2. It introduces a lot of errors, which I think point to valid use cases, which are now nicely written in the tests I think:

```
> unified@10.0.0-beta.1 build
> rimraf "test/**/*.d.ts" && tsc && tsd && type-coverage

lib/index.js:138:23 - error TS2345: Argument of type 'unknown[]' is not assignable to parameter of type '[] | [boolean]'.
  Type 'unknown[]' is not assignable to type '[boolean]'.
    Target requires 1 element(s) but source may have fewer.

138       destination.use(...attachers[index])
                          ~~~~~~~~~~~~~~~~~~~

lib/index.js:193:52 - error TS2556: A spread argument must either have a tuple type or be passed to a rest parameter.

193       const transformer = attacher.call(processor, ...options)
                                                       ~~~~~~~~~~

test/use.js:103:12 - error TS2769: No overload matches this call.
  Overload 1 of 3, '(plugin: Plugin<[]>, ...settings: [] | [boolean]): Processor', gave the following error.
    Argument of type '{}[][]' is not assignable to parameter of type 'Plugin<[]>'.
      Type '{}[][]' provides no match for the signature '(this: Processor): void | Transformer'.

103       .use([
               ~
104         [
    ~~~~~~~~~
... 
123         ]
    ~~~~~~~~~
124       ])
    ~~~~~~~


test/use.js:118:15 - error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.

118               this,
                  ~~~~

  test/use.js:116:11
    116           function () {
                  ~~~~~~~~
    An outer value of 'this' is shadowed by this container.

test/use.js:217:7 - error TS2769: No overload matches this call.
  Overload 1 of 3, '(plugin: Plugin<[]>, ...settings: [] | [boolean]): Processor', gave the following error.
    Argument of type '(boolean | (() => void))[][]' is not assignable to parameter of type 'Plugin<[]>'.
      Type '(boolean | (() => void))[][]' provides no match for the signature '(this: Processor): void | Transformer'.
  Overload 2 of 3, '(tuple: [Plugin<[(boolean | (() => void))[]]>, (boolean | (() => void))[]] | [Plugin<[(boolean | (() => void))[]]>, boolean]): Processor', gave the following error.
    Type '(() => void)[]' is not assignable to type 'Plugin<[(boolean | (() => void))[]]>'.
      Type '(() => void)[]' provides no match for the signature '(this: Processor, settings_0: (boolean | (() => void))[]): void | Transformer'.
  Overload 3 of 3, '(presetOrList: Preset | PluggableList): Processor', gave the following error.
    Type 'false' is not assignable to type 'undefined'.

217       processor.use([[plugin], [plugin, false]]).freeze()
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


test/use.js:234:9 - error TS2769: No overload matches this call.
  Overload 2 of 3, '(tuple: [Plugin<[(boolean | (() => void))[]]>, (boolean | (() => void))[]] | [Plugin<[(boolean | (() => void))[]]>, boolean]): Processor', gave the following error.
    Type '(false | (() => void))[]' is not assignable to type 'Plugin<[(boolean | (() => void))[]]>'.
      Type '(false | (() => void))[]' provides no match for the signature '(this: Processor, settings_0: (boolean | (() => void))[]): void | Transformer'.

234         [plugin, false],
            ~~~~~~~~~~~~~~~


test/use.js:251:9 - error TS2769: No overload matches this call.
  Overload 2 of 3, '(tuple: [Plugin<[(((options?: unknown) => void) | { foo: boolean; })[]]>, (((options?: unknown) => void) | { foo: boolean; })[]] | [Plugin<[(((options?: unknown) => void) | { foo: boolean; })[]]>, boolean]): Processor', gave the following error.
    Type '(false | ((options?: unknown) => void))[]' is not assignable to type 'Plugin<[(((options?: unknown) => void) | { foo: boolean; })[]]>'.
      Type '(false | ((options?: unknown) => void))[]' provides no match for the signature '(this: Processor, settings_0: (((options?: unknown) => void) | { foo: boolean; })[]): void | Transformer'.

251         [plugin, false],
            ~~~~~~~~~~~~~~~


test/use.js:402:9 - error TS2769: No overload matches this call.
  Overload 2 of 3, '(tuple: [Plugin<[]>] | [Plugin<[]>, boolean]): Processor', gave the following error.
    Argument of type '{ plugins: {}[][]; }' is not assignable to parameter of type '[Plugin<[]>] | [Plugin<[]>, boolean]'.
      Object literal may only specify known properties, and 'plugins' does not exist in type '[Plugin<[]>] | [Plugin<[]>, boolean]'.

402         plugins: [
            ~~~~~~~~~~
403           [plugin1, one],
    ~~~~~~~~~~~~~~~~~~~~~~~~~
404           [plugin2, two]
    ~~~~~~~~~~~~~~~~~~~~~~~~
405         ]
    ~~~~~~~~~


test/use.js:434:9 - error TS2769: No overload matches this call.
  Overload 2 of 3, '(tuple: [Plugin<[]>] | [Plugin<[]>, boolean]): Processor', gave the following error.
    Argument of type '{ plugins: { plugins: {}[][]; }[]; }' is not assignable to parameter of type '[Plugin<[]>] | [Plugin<[]>, boolean]'.
      Object literal may only specify known properties, and 'plugins' does not exist in type '[Plugin<[]>] | [Plugin<[]>, boolean]'.

434         plugins: [{plugins: [[plugin1, one]]}, {plugins: [[plugin2, two]]}]
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~



Found 9 errors.
```

---

 I’d say `any[]` is fine, which people can overwrite.